dokka-integration-tests/gradle/build.gradle.kts

```diff
@@ -1,43 +1,208 @@
+    api(libs.kotlin.test)
+    api(libs.junit.jupiterApi)
+    api(libs.junit.jupiterParams)
 
-    implementation(libs.jsoup)
+    api(gradleTestKit())
 }
 
 kotlin {
     // this project only contains test utils and isn't published, so it doesn't matter about explicit API
-    explicitApi = ExplicitApiMode.Disabled
+    explicitApi = Disabled
+
+    compilerOptions {
+        allWarningsAsErrors = false
+        optIn.add("kotlin.io.path.ExperimentalPathApi")
+    }
 }
 
 val aggregatingProject = gradle.includedBuild("dokka")
+val templateProjectsDir = layout.projectDirectory.dir("projects")
 
-tasks.integrationTest {
+tasks.integrationTestPreparation {
+    // TODO remove this in https://github.com/Kotlin/dokka/pull/3433
     dependsOn(aggregatingProject.task(":publishToMavenLocal"))
+}
+
+tasks.withType<Test>().configureEach {
+    dependsOn(tasks.integrationTestPreparation)
+
+    setForkEvery(1)
+    maxHeapSize = "2G"
+    dokkaBuild.integrationTestParallelism.orNull?.let { parallelism ->
+        maxParallelForks = parallelism
+    }
+
+    val useK2 = dokkaBuild.integrationTestUseK2.get()
+
+    useJUnitPlatform {
+        if (useK2) excludeTags("onlyDescriptors", "onlyDescriptorsMPP")
+    }
+
+    systemProperty("org.jetbrains.dokka.experimental.tryK2", useK2)
+    // allow inspecting projects in temporary dirs after a test fails
+    systemProperty(
+        "junit.jupiter.tempdir.cleanup.mode.default",
+        dokkaBuild.isCI.map { isCi -> if (isCi) "ALWAYS" else "ON_SUCCESS" }.get(),
+    )
 
     environment("DOKKA_VERSION", project.version)
+    environment("isExhaustive", dokkaBuild.integrationTestExhaustive)
+    environment("ANDROID_HOME", dokkaBuild.androidSdkDir.get().invariantSeparatorsPath)
+
+    testLogging {
+        exceptionFormat = FULL
+        events(SKIPPED, FAILED)
+        showExceptions = true
+        showCauses = true
+        showStackTraces = true
+    }
+
+    // TODO remove this in https://github.com/Kotlin/dokka/pull/3433
+    doNotTrackState("uses artifacts from Maven Local")
+}
+
+@Suppress("UnstableApiUsage")
+testing {
+    suites {
+        withType<JvmTestSuite>().configureEach {
+            useJUnitJupiter()
+
+            dependencies {
+                implementation(project())
+            }
 
-    inputs.dir(file("projects"))
+            targets.configureEach {
+                testTask.configure {
+                    doFirst {
+                        logger.info("running $path with javaLauncher:${javaLauncher.orNull?.metadata?.javaRuntimeVersion}")
+                    }
+                }
+            }
+        }
+
+        /**
+         * Create a new [JvmTestSuite] for a Gradle project.
+         *
+         * @param[projectPath] path to the Gradle project that will be tested by this suite, relative to [templateProjectsDir].
+         * The directory will be passed as a system property, `templateProjectDir`.
+         */
+        fun registerTestProjectSuite(
```

Comment:
IMO, it would be better to move this function to the bottom of the file (or to `build-logic`) if possible to reduce context-switching while reading the file